## Data Cleaning & Feature Engineering:

In [81]:
# Let's start by importing all our libraries required
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Also importing warnings so we can avoid the red warnings in our results
import warnings
warnings.filterwarnings("ignore")

In [82]:
# Let's import our dataset now
df= pd.read_csv('capstone_data/StartupFile_USA_Final.csv')
df

startupName                homepage_url  \
0             One Inc.          http://whatis1.com   
1          10-20 Media   http://www.10-20media.com   
2                Redox  http://www.redoxengine.com   
3                10sec             https://10s.ec/   
4     10X Technologies  http://10xtechnologies.com   
...                ...                         ...   
9618        Zymr, Inc.         http://www.zymr.com   
9619             Zynga        http://www.zynga.com   
9620            ZYOMYX       http://www.zyomyx.com   
9621              Zype         http://www.zype.com   
9622             Zyrra        http://www.zyrra.com   

                                          category_list  total_funding  \
0                                                mobile      1150050.0   
1                                            e-commerce      2050000.0   
2     health care|health care information technology...      4000000.0   
3            e-commerce|mobile commerce|social commerce      1600000.0   
4                                         biotechnology      3550000.0   
...                                                 ...            ...   
9618  android|apps|cloud computing|cloud infrastruct...       100000.0   
9619  facebook applications|games|networking|technology    866550786.0   
9620                                      biotechnology     34275015.0   
9621  cloud computing|mobile advertising|video|video...      3300000.0   
9622                                         e-commerce      1510500.0   

         status country_code state_code  funding_rounds first_funding_at  \
0     operating          USA         CA               3          7/20/11   
1     operating          USA         MD               4          6/18/09   
2     operating          USA         WI               2         11/26/14   
3     operating          USA         CA               2         10/31/13   
4     operating          USA         CA               2          11/7/12   
...         ...          ...        ...             ...              ...   
9618  operating          USA         CA               3         12/12/12   
9619        ipo          USA         CA               9           1/1/07   
9620  operating          USA         MO               4          1/29/10   
9621  operating          USA         NY               2           4/1/15   
9622  operating          USA         MA               4         10/21/10   

     last_funding_at  homepage_url_code     time_delayFor_first_round  \
0             2/5/14                  1   566 days 00:00:00.000000000   
1           12/28/11                  1   285 days 00:00:00.000000000   
2           10/20/15                  1   328 days 00:00:00.000000000   
3             5/8/14                  1   189 days 00:00:00.000000000   
4            1/12/15                  1   796 days 00:00:00.000000000   
...              ...                ...                           ...   
9618          1/1/14                  1    20 days 00:00:00.000000000   
9619         2/18/11                  1   365 days 00:00:00.000000000   
9620         3/26/14                  1  1248 days 00:00:00.000000000   
9621        10/27/15                  1   209 days 00:00:00.000000000   
9622        10/18/12                  1    25 days 00:00:00.000000000   

        average_time_delay_rounds  average_raised_usd  \
0     465 days 12:00:00.000000000        5.750250e+05   
1     307 days 16:00:00.000000000        5.125000e+05   
2     328 days 00:00:00.000000000        2.000000e+06   
3     189 days 00:00:00.000000000        1.600000e+06   
4     796 days 00:00:00.000000000        1.775000e+06   
...                           ...                 ...   
9618  192 days 12:00:00.000000000        2.394698e+07   
9619  188 days 15:00:00.000000000        1.083188e+08   
9620  505 days 16:00:00.000000000        8.568754e+06   
9621  209 days 00:00:00.000000000        1.650000e+06   
9622  242 days 16:00:00.000000000        3.776250e+05   

In [83]:
# Let's check how our data-set looks
df.describe()

total_funding  funding_rounds  homepage_url_code  average_raised_usd  \
count   9.623000e+03     9623.000000        9623.000000        9.623000e+03   
mean    3.063658e+07        3.277772           0.975579        8.919254e+06   
std     3.143715e+08        1.747181           0.154359        6.385372e+07   
min     6.513000e+03        1.000000           0.000000        0.000000e+00   
25%     2.000000e+06        2.000000           1.000000        9.179248e+05   
50%     7.595758e+06        3.000000           1.000000        3.150000e+06   
75%     2.627141e+07        4.000000           1.000000        9.388061e+06   
max     3.007950e+10       19.000000           1.000000        6.015901e+09   

       average_time_delay_rounds_number  time_delayFor_first_round_number  \
count                       9623.000000                       9623.000000   
mean                         492.282240                        505.329939   
std                          382.902718                        442.626397   
min                            1.000000                          0.000000   
25%                          256.000000                        210.000000   
50%                          399.000000                        391.000000   
75%                          610.000000                        670.000000   
max                         4081.000000                       5159.000000   

        statusCode  No_Of_Investors  
count  9623.000000      9623.000000  
mean      0.748623         4.002494  
std       0.433827         4.354342  
min       0.000000         1.000000  
25%       0.000000         1.000000  
50%       1.000000         2.000000  
75%       1.000000         5.000000  
max       1.000000        50.000000

In [84]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9623 entries, 0 to 9622
Data columns (total 20 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   startupName                       9623 non-null   object 
 1   homepage_url                      9388 non-null   object 
 2   category_list                     9623 non-null   object 
 3   total_funding                     9623 non-null   float64
 4   status                            9623 non-null   object 
 5   country_code                      9623 non-null   object 
 6   state_code                        9623 non-null   object 
 7   funding_rounds                    9623 non-null   int64  
 8   first_funding_at                  9623 non-null   object 
 9   last_funding_at                   9623 non-null   object 
 10  homepage_url_code                 9623 non-null   int64  
 11  time_delayFor_first_round         9623 non-null   object 
 12  averag

In [85]:
# Let's check if we have any nulls
df.isna().sum()

startupName                           0
homepage_url                        235
category_list                         0
total_funding                         0
status                                0
country_code                          0
state_code                            0
funding_rounds                        0
first_funding_at                      0
last_funding_at                       0
homepage_url_code                     0
time_delayFor_first_round             0
average_time_delay_rounds             0
average_raised_usd                    0
average_time_delay_rounds_number      0
time_delayFor_first_round_number      0
statusCode                            0
No_Of_Investors                       0
MainCategory                          0
USA_Region                            0
dtype: int64

Only the 'homepage_url' column has any nulls. Let's try and drop columns we don't need and feature engineer some new ones.

In [86]:
df['homepage_url'].duplicated().sum()

236

Here we can see that we have 236 duplicates let's take a look and see if there's something we can do.

In [87]:
df[df['homepage_url'].duplicated()]

startupName homepage_url  \
101          Accelerated Pharma          NaN   
110             Accept Software          NaN   
227             ADVANCE Medical          NaN   
245     Advanced System Designs          NaN   
336                       AkaRx          NaN   
...                         ...          ...   
9221  Waterford Battery Systems          NaN   
9304                        Whi          NaN   
9335                Windtronics          NaN   
9421                  XenSource          NaN   
9551         ZetaRx Biosciences          NaN   

                                        category_list  total_funding  \
101                                   pharmaceuticals      2549999.0   
110                        career management|software     25305000.0   
227                                       health care     26000000.0   
245                                          software      1325000.0   
336                                     biotechnology     26100000.0   
...                                               ...            ...   
9221  batteries|electrical distribution|manufacturing      1594130.0   
9304          health and wellness|health care|medical     11317850.0   
9335                              hardware + software      4000000.0   
9421                                         software     38000000.0   
9551                                    biotechnology      2750000.0   

         status country_code state_code  funding_rounds first_funding_at  \
101   operating          USA         IL               2           6/4/15   
110    acquired          USA         CA               4          9/16/05   
227   operating          USA         CA               2          5/31/11   
245   operating          USA         IL               2           6/9/09   
336    acquired          USA         NJ               5          8/22/05   
...         ...          ...        ...             ...              ...   
9221  operating          USA         FL               2          3/17/14   
9304  operating          USA         CA               2           4/2/14   
9335  operating          USA         MI               2           4/1/09   
9421  operating          USA         CA               3          1/13/05   
9551  operating          USA         WA               2         12/22/10   

     last_funding_at  homepage_url_code    time_delayFor_first_round  \
101         11/23/15                  0  172 days 00:00:00.000000000   
110         12/14/10                  0  314 days 00:00:00.000000000   
227           9/4/13                  0  827 days 00:00:00.000000000   
245          8/17/10                  0  434 days 00:00:00.000000000   
336          7/13/07                  0  261 days 00:00:00.000000000   
...              ...                ...                          ...   
9221         7/15/15                  0  485 days 00:00:00.000000000   
9304         4/13/15                  0  376 days 00:00:00.000000000   
9335         1/12/10                  0  286 days 00:00:00.000000000   
9421        10/31/06                  0  231 days 00:00:00.000000000   
9551        12/12/12                  0  721 days 00:00:00.000000000   

        average_time_delay_rounds  average_raised_usd  \
101   172 days 00:00:00.000000000          1274999.50   
110   638 days 08:00:00.000000000          6326250.00   
227   827 days 00:00:00.000000000         13000000.00   
245   434 days 00:00:00.000000000           662500.00   
336   172 days 12:00:00.000000000          5220000.00   
...                           ...                 ...   
9221  485 days 00:00:00.000000000           797065.00   
9304  376 days 00:00:00.000000000          5658925.00   
9335  286 days 00:00:00.000000000          2000000.00   
9421  328 days 00:00:00.000000000         12666666.67   
9551  721 days 00:00:00.000000000          1375000.00   

      average_time_delay_rounds_number  time_delayFor_first_round_number  \
101                                172 

In [88]:
df['homepage_url'].isna().sum()

235

Since 235 of the duplicates have null values in them, we can input them as a new feature and make an assumption that the start-ups which have nulls in their homepage url have no website currently available.

In [89]:
df['website_availibility']= df['homepage_url'].notnull().astype('int')

The **'startupName'** column has names of all the start-ups we cannot convert all the names to numeric so we can drop them since we won't be using them.\
We can drop the **'homepage_url'** since we changed it into a feature.\
Since **'category_list'** shows the categories and 'MainCategory' has the specified category of each start-up, we can drop the 'category_list' too.\
We don't have much information about **'homepage_url_code'** and we already created a variable for website availibility so we can drop it too.\
The **'statusCode'** can also be dropped since we created a column of the 'status' for our start-ups.\
The following columns **'time_delayFor_first_round','average_time_delay_rounds','first_funding_at', 'last_funding_at'** seem duplicates since we have other columns resembling them so it's safe to drop them as well.\
Since we have decided to choose our country as U.S all the rows belong to U.S and the column **'country_code'** is not of any importance so we can drop it as well.\
Let's drop all our columns now:


In [90]:
df.drop(['startupName','homepage_url','category_list','homepage_url_code',
          'statusCode','time_delayFor_first_round','average_time_delay_rounds', 'first_funding_at', 
         'last_funding_at','country_code'],
          axis=1, inplace=True)
df

total_funding     status state_code  funding_rounds  average_raised_usd  \
0         1150050.0  operating         CA               3        5.750250e+05   
1         2050000.0  operating         MD               4        5.125000e+05   
2         4000000.0  operating         WI               2        2.000000e+06   
3         1600000.0  operating         CA               2        1.600000e+06   
4         3550000.0  operating         CA               2        1.775000e+06   
...             ...        ...        ...             ...                 ...   
9618       100000.0  operating         CA               3        2.394698e+07   
9619    866550786.0        ipo         CA               9        1.083188e+08   
9620     34275015.0  operating         MO               4        8.568754e+06   
9621      3300000.0  operating         NY               2        1.650000e+06   
9622      1510500.0  operating         MA               4        3.776250e+05   

      average_time_delay_rounds_number  time_delayFor_first_round_number  \
0                                  465                               566   
1                                  307                               285   
2                                  328                               328   
3                                  189                               189   
4                                  796                               796   
...                                ...                               ...   
9618                               192                                20   
9619                               188                               365   
9620                               505                              1248   
9621                               209                               209   
9622                               242                                25   

      No_Of_Investors MainCategory USA_Region  website_availibility  
0                  12       mobile       West                     1  
1                   1   e-commerce      South                     1  
2                   6   healthcare    Midwest                     1  
3                   2       mobile       West                     1  
4                   1      biotech       West                     1  
...               ...          ...        ...                   ...  
9618                1       mobile       West                     1  
9619               22        games       West                     1  
9620                3      biotech    Midwest                     1  
9621               10       mobile  Northeast                     1  
9622                2   e-commerce  Northeast                     1  

[9623 rows x 11 columns]

#### Since the 'status' column is our target column let's check if it needs any feature engineering

In [91]:
df['status'].value_counts()

operating    7495
acquired     1160
closed        521
ipo           447
Name: status, dtype: int64

Here we have 4 classes in our 'status' column, so we can reduce these classes into a 2 class binary classification problem.
After researching it was found that the **'ipo'** class states that the start-up has filed for debt, so we can change that class to closed.\
The 'acquiring' class is the start-up that has been recently acquired so we can assume this to be in the operating class.
To summarize we are going to put the **'operating'** and the **'acquired'** classes to be a binary **1** i.e the operating class or **successful**\
while the **'closed'** and the **'ipo'** will be under binary class **0** i.e closed or **failed** start-ups.

In [92]:
#Making a copy if we need it
data= df.copy()

In [93]:
# Let's binarize our classes
df['status'].replace({"acquired":1 ,"operating":1, "ipo":0, "closed":0},inplace=True)

In [94]:
df['status'].value_counts()

1    8655
0     968
Name: status, dtype: int64

After checking the value counts we can see that the we have two classes '1' and '0' now.

Let's check in our other columns and see if we can create some features.

In [95]:
df['MainCategory'].value_counts()

software         3050
mobile           1514
biotech          1454
healthcare       1282
e-commerce        657
games             483
technology        447
pharma            302
manufacturing     229
medicalGear       205
Name: MainCategory, dtype: int64

In [96]:
df['USA_Region'].value_counts()

West         4673
Northeast    2289
South        1678
Midwest       983
Name: USA_Region, dtype: int64

In [97]:
df['state_code'].unique()

array(['CA', 'MD', 'WI', 'RI', 'MA', 'NY', 'DC', 'IN', 'OH', 'IL', 'WA',
       'GA', 'MI', 'FL', 'MN', 'SC', 'CT', 'KY', 'VA', 'PA', 'NC', 'TX',
       'CO', 'KS', 'TN', 'AZ', 'OK', 'NJ', 'OR', 'MO', 'AR', 'UT', 'NH',
       'LA', 'NM', 'SD', 'MT', 'NE', 'NV', 'blank', 'DE', 'ID', 'IA',
       'VT', 'MS', 'AK', 'AL', 'HI', 'ND', 'ME', 'WY', 'WV'], dtype=object)

In [98]:
df.loc[df['state_code']== 'blank']

total_funding  status state_code  funding_rounds  average_raised_usd  \
469    3.740000e+08       0      blank               4        9.350000e+07   
3135   5.000000e+05       0      blank               2        2.500000e+05   
3833   4.700000e+05       1      blank               2        2.350000e+05   
4932   4.420632e+05       1      blank               3        1.473543e+05   
8792   1.175000e+06       1      blank               3        3.916667e+05   

      average_time_delay_rounds_number  time_delayFor_first_round_number  \
469                                196                                85   
3135                               421                               421   
3833                                92                                92   
4932                               396                               274   
8792                               206                               237   

      No_Of_Investors   MainCategory USA_Region  website_availibility  
469                15         mobile       West                     1  
3135                1  manufacturing       West                     0  
3833                2         mobile       West                     1  
4932                2         mobile       West                     1  
8792                1         pharma       West                     0

Since we have some 'blank' values in our state code, let's drop them to avoid any confusion later.

In [99]:
df.drop(df.loc[df['state_code']=='blank'].index, inplace=True)

In [100]:
df['state_code'].unique()

array(['CA', 'MD', 'WI', 'RI', 'MA', 'NY', 'DC', 'IN', 'OH', 'IL', 'WA',
       'GA', 'MI', 'FL', 'MN', 'SC', 'CT', 'KY', 'VA', 'PA', 'NC', 'TX',
       'CO', 'KS', 'TN', 'AZ', 'OK', 'NJ', 'OR', 'MO', 'AR', 'UT', 'NH',
       'LA', 'NM', 'SD', 'MT', 'NE', 'NV', 'DE', 'ID', 'IA', 'VT', 'MS',
       'AK', 'AL', 'HI', 'ND', 'ME', 'WY', 'WV'], dtype=object)

Now, we don't have any 'blank' values in our state_code.

In [101]:
data.shape

(9623, 11)

Since we don't have many categories in the 'MainCategory' and 'USA_Region' we can convert them into dummies.\
For the states since we have a lot of states, we can convert it into numbers with Label Encoder.

In [102]:
from sklearn.preprocessing import LabelEncoder

labelencoder=LabelEncoder()
df['state_code']=labelencoder.fit_transform(df['state_code'])

df= pd.get_dummies(df,columns=['MainCategory','USA_Region'], drop_first= True)

In [103]:
df

total_funding  status  state_code  funding_rounds  average_raised_usd  \
0         1150050.0       1           4               3        5.750250e+05   
1         2050000.0       1          20               4        5.125000e+05   
2         4000000.0       1          48               2        2.000000e+06   
3         1600000.0       1           4               2        1.600000e+06   
4         3550000.0       1           4               2        1.775000e+06   
...             ...     ...         ...             ...                 ...   
9618       100000.0       1           4               3        2.394698e+07   
9619    866550786.0       0           4               9        1.083188e+08   
9620     34275015.0       1          24               4        8.568754e+06   
9621      3300000.0       1          34               2        1.650000e+06   
9622      1510500.0       1          19               4        3.776250e+05   

      average_time_delay_rounds_number  time_delayFor_first_round_number  \
0                                  465                               566   
1                                  307                               285   
2                                  328                               328   
3                                  189                               189   
4                                  796                               796   
...                                ...                               ...   
9618                               192                                20   
9619                               188                               365   
9620                               505                              1248   
9621                               209                               209   
9622                               242                                25   

      No_Of_Investors  website_availibility  MainCategory_e-commerce  ...  \
0                  12                     1                        0  ...   
1                   1                     1                        1  ...   
2                   6                     1                        0  ...   
3                   2                     1                        0  ...   
4                   1                     1                        0  ...   
...               ...                   ...                      ...  ...   
9618                1                     1                        0  ...   
9619               22                     1                        0  ...   
9620                3                     1                        0  ...   
9621               10                     1                        0  ...   
9622                2                     1                        1  ...   

      MainCategory_healthcare  MainCategory_manufacturing  \
0                           0                           0   
1                           0                           0   
2                           1                           0   
3                           0                           0   
4                           0                           0   
...                       ...                         ...   
9618                        0                           0   
9619                        0                           0   
9620                        0                           0   
9621                        0                           0   
9622                        0                           0   

      MainCategory_medicalGear  MainCategory_mobile  MainCategory_pharma  \
0                            0                    1                    0   
1                            0                    0                    0   
2                            0                    0                    0   
3                            0                    1                    0   
4                            0                    0                    0   
...                        ...

In [106]:
df['average_raised_usd'].dtype

dtype('float64')

Since the 'average_raised_usd' is in the float data type and in scientific notation let's convert it into an integer data type.

In [109]:
df['average_raised_usd']= df['average_raised_usd'].astype(int)

In [110]:
df['average_raised_usd'].dtype

dtype('int32')

In [111]:
df

total_funding  status  state_code  funding_rounds  average_raised_usd  \
0         1150050.0       1           4               3              575025   
1         2050000.0       1          20               4              512500   
2         4000000.0       1          48               2             2000000   
3         1600000.0       1           4               2             1600000   
4         3550000.0       1           4               2             1775000   
...             ...     ...         ...             ...                 ...   
9618       100000.0       1           4               3            23946976   
9619    866550786.0       0           4               9           108318848   
9620     34275015.0       1          24               4             8568753   
9621      3300000.0       1          34               2             1650000   
9622      1510500.0       1          19               4              377625   

      average_time_delay_rounds_number  time_delayFor_first_round_number  \
0                                  465                               566   
1                                  307                               285   
2                                  328                               328   
3                                  189                               189   
4                                  796                               796   
...                                ...                               ...   
9618                               192                                20   
9619                               188                               365   
9620                               505                              1248   
9621                               209                               209   
9622                               242                                25   

      No_Of_Investors  website_availibility  MainCategory_e-commerce  ...  \
0                  12                     1                        0  ...   
1                   1                     1                        1  ...   
2                   6                     1                        0  ...   
3                   2                     1                        0  ...   
4                   1                     1                        0  ...   
...               ...                   ...                      ...  ...   
9618                1                     1                        0  ...   
9619               22                     1                        0  ...   
9620                3                     1                        0  ...   
9621               10                     1                        0  ...   
9622                2                     1                        1  ...   

      MainCategory_healthcare  MainCategory_manufacturing  \
0                           0                           0   
1                           0                           0   
2                           1                           0   
3                           0                           0   
4                           0                           0   
...                       ...                         ...   
9618                        0                           0   
9619                        0                           0   
9620                        0                           0   
9621                        0                           0   
9622                        0                           0   

      MainCategory_medicalGear  MainCategory_mobile  MainCategory_pharma  \
0                            0                    1                    0   
1                            0                    0                    0   
2                            0                    0                    0   
3                            0                    1                    0   
4                            0                    0                    0   
...                        ...

Let's check for duplicates:

In [107]:
df.duplicated().sum()

1

Since we only one duplicate row we can drop it.

In [116]:
df.drop_duplicates(inplace= True)

In [117]:
#Re-checking
df.duplicated().sum()

0

### Now since we have cleared and formatted our data, we can save the data file and use it for modelling and EDA.

In [120]:
df.to_csv('Capstone data/startup_dataset.csv')

In [119]:
# import joblib
# joblib.dump(df, 'Capstone data/startup_dataset.csv')

['Capstone data/startup_dataset.csv']